staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -94,45 +96,54 @@ func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
-	config, err := loadConfig(filepath)
+// LoadEncryptionConfig parses encryption configuration file and returns:
+// - prefix transformers
+// - health checks
+// - config file content hash
+// - max timeout among all transformers
+func LoadEncryptionConfig(filepath string, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, string, time.Duration, error) {
+	config, contentHash, err := loadConfig(filepath)
 	if err != nil {
-		return nil, nil, fmt.Errorf("error while parsing file: %v", err)
+		return nil, nil, contentHash, 0, fmt.Errorf("error while parsing file: %v", err)
 	}
 
-	return getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
+	transformers, healthChecks, kmsMaxTimeout, err := getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
+
+	return transformers, healthChecks, contentHash, kmsMaxTimeout, err
 }
 
-func getTransformerOverridesAndKMSPluginHealthzCheckers(config *apiserverconfig.EncryptionConfiguration, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, error) {
+func getTransformerOverridesAndKMSPluginHealthzCheckers(config *apiserverconfig.EncryptionConfiguration, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, time.Duration, error) {
 	var kmsHealthChecks []healthz.HealthChecker
-	transformers, probes, err := getTransformerOverridesAndKMSPluginProbes(config, stopCh)
+	transformers, probes, kmsMaxTimeout, err := getTransformerOverridesAndKMSPluginProbes(config, stopCh)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, kmsMaxTimeout, err
 	}
 	for i := range probes {
 		probe := probes[i]
 		kmsHealthChecks = append(kmsHealthChecks, probe.toHealthzCheck(i))
 	}
 
-	return transformers, kmsHealthChecks, nil
+	return transformers, kmsHealthChecks, kmsMaxTimeout, nil
 }
 
 type healthChecker interface {
 	toHealthzCheck(idx int) healthz.HealthChecker
 }
 
-func getTransformerOverridesAndKMSPluginProbes(config *apiserverconfig.EncryptionConfiguration, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthChecker, error) {
+func getTransformerOverridesAndKMSPluginProbes(config *apiserverconfig.EncryptionConfiguration, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthChecker, time.Duration, error) {
 	resourceToPrefixTransformer := map[schema.GroupResource][]value.PrefixTransformer{}
 	var probes []healthChecker
+	var kmsMaxTimeout time.Duration
 
 	// For each entry in the configuration
 	for _, resourceConfig := range config.Resources {
 		resourceConfig := resourceConfig
 
-		transformers, p, err := prefixTransformersAndProbes(resourceConfig, stopCh)
+		transformers, p, timeout, err := prefixTransformersAndProbes(resourceConfig, stopCh)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, timeout, err
 		}
+		kmsMaxTimeout = timeout
```

Comment:
Where do you compare the values to get the max timeout? The current implementation will always return the last timeout as max.

Reply:
I compare it here: https://github.com/kubernetes/kubernetes/pull/112050/files#diff-8ca5bf5da3d263c7a4d934647ad0f3d929256c9b399d305576b0fc84c6b966deR303-R306

Unit tests seem to agree with this:
https://github.com/kubernetes/kubernetes/pull/112050/files#diff-f2f067d6aa03c572d14285964d389066c885f5f54df7dc7d779ccecdc50f7b75R259

Reply:
prefixTransformersAndProbes returns the max timeout from one resourceConfig's providers. where do you compare the max timeout for all resourceConfigs?

Reply:
Ah, good catch. Updated the logic and tests. Thanks, @ritazh!